stan/math/rev/core/vari.hpp

```diff
@@ -68,37 +117,156 @@ class vari {
+   *
+   * This no-op implementation enables a subclass to throw
+   * exceptions in its constructor.  An exception thrown in the
+   * constructor of a subclass will result in an error being
+   * raised, which is in turn caught and calls delete().
+   *
+   * See the discussion of "plugging the memory leak" in:
+   *   http://www.parashift.com/c++-faq/memory-pools.html
+   */
+  static inline void operator delete(void* /* ignore arg */) { /* no op */
+  }
+};
+
+template <typename T>
+class vari_value<T, std::enable_if_t<is_eigen<T>::value>> : public vari_base {
+ private:
+  template <typename>
+  friend class var_value;
+  using eigen_scalar = value_type_t<T>;
+  value_type_t<T>* val_mem_;
+  value_type_t<T>* adj_mem_;
+
+ public:
+  using Scalar = T;
+  using value_type = T;
+  using PlainObject = plain_type_t<T>;
+  using eigen_map = Eigen::Map<PlainObject>;
+  const Eigen::Index rows_;
+  const Eigen::Index cols_;
+  const Eigen::Index size_;
+  const Eigen::Index RowsAtCompileTime = T::RowsAtCompileTime;
+  const Eigen::Index ColsAtCompileTime = T::ColsAtCompileTime;
+  /**
+   * The value of this variable.
+   */
+  eigen_map val_;
+
+  /**
+   * The adjoint of this variable, which is the partial derivative
+   * of this variable with respect to the root variable.
+   */
+  eigen_map adj_;
+
+  /**
+   * Construct a variable implementation from a value.  The
+   * adjoint is initialized to zero.
+   *
+   * All constructed variables are added to the stack.  Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @param x Value of the constructed variable.
+   */
+  explicit vari_value(const T& x)
+      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        val_(val_mem_, x.rows(), x.cols()),
+        adj_(adj_mem_, x.rows(), x.cols()),
+        rows_(x.rows()),
+        cols_(x.cols()),
+        size_(x.size()) {
+    val_ = x;
+    adj_.setZero();
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+  vari_value(const T& x, bool stacked)
+      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        val_(val_mem_, x.rows(), x.cols()),
+        adj_(adj_mem_, x.rows(), x.cols()),
+        rows_(x.rows()),
+        cols_(x.cols()),
+        size_(x.size()) {
+    val_ = x;
+    adj_.setZero();
+    if (stacked) {
+      ChainableStack::instance_->var_stack_.push_back(this);
+    } else {
+      ChainableStack::instance_->var_nochain_stack_.push_back(this);
+    }
+  }
```

Comment:
Could do something similar here with the default argument:

```suggestion
  explicit vari_value(const T& x, bool stacked = 1)
      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
            x.size())),
        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
            x.size())),
        val_(val_mem_, x.rows(), x.cols()),
        adj_(adj_mem_, x.rows(), x.cols()),
        rows_(x.rows()),
        cols_(x.cols()),
        size_(x.size()) {
    val_ = x;
    adj_.setZero();
    if (stacked) {
      ChainableStack::instance_->var_stack_.push_back(this);
    } else {
      ChainableStack::instance_->var_nochain_stack_.push_back(this);
    }
  }
```